### PR TITLE
SOL-151442: add in-memory OAuth token cache with exchanger wiring

### PR DESCRIPTION
**Pull request by amitmorade**
## Summary

Adds an in-memory OAuth token cache (Otter-backed, W-TinyLFU eviction) and wires it into the token exchange flow so that repeated SEMP calls for the same user reuse a cached broker token instead of hitting the IdP on every request.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Performance improvement

## Motivation and Context

Without caching, every SEMP request triggers a full token exchange round-trip to the IdP (~50ms). Composite tools like `get-queue-metrics` make multiple SEMP calls per invocation, multiplying the overhead. The cache eliminates redundant exchanges — observed speedup is ~4.5× on warm hits (310ms → 68ms for a typical tool call).

## Changes Made

- **`internal/oauth/cache/cache.go`** — `TokenCache` interface with structured result types (`GetHit`, `GetMissAbsent`, `GetMissExpired`, `PutStored`, `PutDroppedTTL`, `PutDroppedFull`), `CachedCredential` type with `slog.LogValuer`/`fmt.Stringer` safety guards, and `CacheConfig` struct
- **`internal/oauth/cache/otter.go`** — Otter v2-backed implementation with TTL derived from token `exp` minus clock skew, capped at MaxTTL; `Close()` releases the eviction goroutines on shutdown
- **`internal/oauth/cache/factory.go`** — `NewTokenCache` constructor with MaxSize validation
- **`internal/tokenexchange/exchanger.go`** — Cache wired as mandatory dependency; Get runs before singleflight, Put runs after successful exchange; `Invalidate()` method for cache eviction; `Close()` delegates to cache
- **`internal/semp/auth/oauth.go`** — `HandleAuthFailure` evicts cached token via `Exchanger.Invalidate()` on broker 401, returns `true` to signal retry
- **`cmd/server/main.go`** — Cache instantiated as process-singleton, injected into Exchanger at startup; graceful shutdown calls `Exchanger.Close()` via `defer`
- Cache status types own logging semantics via `String()` and `Level()` methods — the exchanger logs uniformly with zero cache domain knowledge

## Testing

**Test Environment:**
- Go version: 1.25
- OS: macOS / Linux (CI)
- Broker version: 10.x (local Docker)

**Tests Performed:**
- [x] Unit tests added/updated
- [x] Integration tests added/updated
- [x] E2E tests added/updated
- [x] Tested locally with `go test -race ./...`
- [x] Tested with real Solace broker

**Test Coverage:**
- `internal/oauth/cache/` — hit/miss/expiry/eviction/key isolation/structured results/credential safety
- `internal/tokenexchange/exchange_test.go` — cache hit short-circuits IdP, cache miss stores result, `Invalidate` forces refetch, concurrent misses collapse to one IdP call + one Put (singleflight × cache)
- `internal/tokenexchange/exchanger_test.go` — cache nil rejection at construction
- `internal/semp/auth/oauth_test.go` — HandleAuthFailure: no-subject-token returns false; with-subject-token evicts and returns true
- `test/integration/oauth_cache_wiring_test.go` — end-to-end: two tool calls with the same bearer produce exactly one IdP hit and both broker requests carry the identical downstream token
- End-to-end verified against local Solace broker with Keycloak IdP: cold miss → exchange → store → warm hit → skip IdP

## Known limitation (deferred to a followup)

The 401-retry-refresh path is NOT covered end-to-end in this PR. `resilience/sender.go` calls `Authenticator.AddAuth` once before entering `retryablehttp`'s retry loop, so on a broker 401 the retry re-sends the same stale token even though `Exchanger.Invalidate()` has already evicted the cache entry. Under the current wiring, the eviction only helps the NEXT tool call, not the in-flight one.

Fix requires wiring `retryablehttp.PrepareRetry` to re-invoke `AddAuth` before each retry. Tracked as a followup; the failing integration test that demonstrates the bug is drafted but not committed here.

## Breaking Changes

None. The cache is additive — existing behavior is preserved. `Exchanger.New()` now requires a non-nil `TokenCache`, but all callers go through `main.go` which constructs one.

### Code Quality
- [x] Code follows the coding standards
- [x] Self-review completed
- [x] Complex logic has explanatory comments
- [x] No commented-out code or debug statements left in

### Security
- [x] **No credentials in code**
- [x] Followed secure logging rules
- [x] Credential-carrying types implement `slog.LogValuer` — `CachedCredential` redacts `Value` in `String()`, `GoString()`, and `LogValue()`
- [x] No security vulnerabilities introduced

### Testing
- [x] All tests pass locally: `go test -race ./...`
- [x] No race conditions detected
- [x] Edge cases covered by tests
- [x] Error paths tested
- [x] Test names clearly describe what is being tested

### Documentation
- [ ] README.md updated (if user-facing changes)
- [ ] docs/ updated (if architecture/design changes)
- [x] godoc comments added/updated for exported functions

### Git & CI
- [x] Commits have clear, descriptive messages
- [x] Branch is up to date with main
- [x] No merge conflicts
- [x] CI checks passing (lint, build, test, E2E)

## Additional Notes

- This PR supersedes #146. That branch accumulated 21 commits including duplicates re-authored by a stale rebase; this branch is a clean rebase containing only the 15 real SOL-151442 / SOL-150800 commits. File diff vs main is byte-identical (1679 lines) — only the commit list is cleaner.
- Cache defaults: 10,000 max entries, 30s clock skew, 24h max TTL
- Cache is process-singleton — one cache shared across all brokers, keyed by `(subjectToken, brokerAlias)`
- Eviction on 401 is bounded: retry layer allows exactly one retry per SEMP request (`auth401Retried` flag)

## Related Issues/PRs

- Supersedes #146
- Related to SOL-151442
- Depends on #138 (token exchanger wiring)

---

**For Reviewers:**

**Focus Areas**: Cache key construction, TTL calculation (clock skew handling), `HandleAuthFailure` eviction path, credential safety in `CachedCredential`, `Close()` ordering in `main.go`

**Questions**: Should cache size/skew/maxTTL be configurable via `broker-config.yaml` in this PR or a follow-up?

### CODE REVIEW

**Comment by amitmorade** [comment]
Looking into copilot feedback

File: internal/tokenexchange/exchange.go [hunk 1]
```diff
@@ -57,10 +75,45 @@ func (e *Exchanger) Exchange(ctx context.Context, input ExchangeInput) (*Token,
...
+	// Store the exchanged token for future requests.
+	pr, putErr := e.cache.Put(ctx, key, cache.CachedCredential{
+		Value:     tok.Value,
+		ExpiresAt: tok.ExpiresAt,
+	})
+	if putErr != nil {
+		slog.WarnContext(ctx, "token cache put failed", "broker", input.BrokerAlias, "error", putErr)
+	} else {
+		slog.Log(ctx, pr.Status.Level(), "token cache put", "broker", input.BrokerAlias, "status", pr.Status)
+	}
```
[the remark this answers is not in the document]
Comment (amitmorade) [reply internal/tokenexchange/exchange.go]: Good catch — you're right that Put shouldn't run per-waiter. Addressed in [`749f259`](https://github.com/SolaceDev/solace-broker-mcp/pull/150/commits/749f259), but with a different structural fix than the suggested `if !shared` gate. Why not `!shared`:

`singleflight.Group.Do` returns `shared=true` for **every caller that received a value computed elsewhere** — including the winner if any concurrent waiter attached to its in-flight call. So `if !shared { Put(...) }` would skip Put on the winner too whenever there's any concurrent traffic on that key — which is exactly the case we care about. Verified this experimentally: my first attempt used the `!shared` gate and the follow-up `cache.Get` came back `GetMissAbsent`.

Instead: move the Put and its Put-logging **inside** the singleflight func. `Do` runs the func exactly once per burst (only the winner executes; all waiters receive its return value), so Put runs exactly once by construction. The "token exchange succeeded" per-caller latency log stays outside — it's per-caller observability (a caller that cancelled mid-exchange should not see the success log attributed to their request), while Put is per-exchange. The Put uses the detached `exchCtx` — matching the singleflight-detach pattern for the IdP call itself.

Verified with a new `TestExchange_SingleflightWinnerOnlyWritesCache` (in [`749f259`](https://github.com/SolaceDev/solace-broker-mcp/pull/150/commits/749f259)) that wraps the real `TokenCache` in a counting decorator and asserts exactly 1 Put across 100 goroutines bursting on the same key.

File: internal/tokenexchange/exchange.go [hunk 2]
```diff
@@ -19,17 +19,35 @@ import (
...
+				Value:     gr.Entry.Value,
+				ExpiresAt: gr.Entry.ExpiresAt,
+			}, nil
+		}
+	}
+
+	// In-flight dedup: collapse concurrent misses into one IdP call.
 	start := e.nowFunc()
 	v, err, _ := e.group.Do(key, func() (interface{}, error) {
 		// Detach from the caller's context so one caller's cancellation
```
[the remark this answers is not in the document]
Comment (amitmorade) [reply internal/tokenexchange/exchange.go]: Good catch — you're right that Put shouldn't run per-waiter. Addressed in [`749f259`](https://github.com/SolaceDev/solace-broker-mcp/pull/150/commits/749f259), but with a different structural fix than the suggested `if !shared` gate. Why not `!shared`:

`singleflight.Group.Do` returns `shared=true` for **every caller that received a value computed elsewhere** — including the winner if any concurrent waiter attached to its in-flight call. So `if !shared { Put(...) }` would skip Put on the winner too whenever there's any concurrent traffic on that key — which is exactly the case we care about. Verified this experimentally: my first attempt used the `!shared` gate and the follow-up `cache.Get` came back `GetMissAbsent`.

Instead: move the Put and its Put-logging **inside** the singleflight func. `Do` runs the func exactly once per burst (only the winner executes; all waiters receive its return value), so Put runs exactly once by construction. The "token exchange succeeded" per-caller latency log stays outside — it's per-caller observability (a caller that cancelled mid-exchange should not see the success log attributed to their request), while Put is per-exchange. The Put uses the detached `exchCtx` — matching the singleflight-detach pattern for the IdP call itself.

Verified with a new `TestExchange_SingleflightWinnerOnlyWritesCache` (in [`749f259`](https://github.com/SolaceDev/solace-broker-mcp/pull/150/commits/749f259)) that wraps the real `TokenCache` in a counting decorator and asserts exactly 1 Put across 100 goroutines bursting on the same key.

**Review by solace-aross — COMMENTED** [review]
## Review

Solid core design — SHA-256 keyed dedup with no cross-user leak risk, cache failures never block a real exchange, memory bounds are sane, and rollback is trivial (pure in-memory, no migration). Copilot already caught most of the mechanical issues below; noting where I agree/prioritize and adding a couple it missed.

### Worth fixing before merge

**1. `HandleAuthFailure` returning `true` makes the 401 path worse than before it existed** — `internal/semp/auth/oauth.go:88`

Eviction happens unconditionally now, but `resilience/sender.go` never re-invokes `AddAuth` before the retry (the limitation you've already documented, SOL-151624), so the retry always replays the identical stale header and 401s again. Versus the old `return false`: same eventual failure, plus one wasted round-trip + backoff, plus a `"auth handler recovered"` WARN log line that says the opposite of what happened — which will read as a confusing false signal to whoever's debugging a slow/failing OAuth broker call. Documenting the gap is good; shipping behavior that's strictly worse than what it replaces for a full release cycle isn't. Would keeping the eviction but returning `false` until the `PrepareRetry` wiring lands be an option, so this ships without the regression?

**2. Singleflight `shared` is discarded — every waiter redundantly writes and logs** — `internal/tokenexchange/exchange.go` (`v, err, _ := e.group.Do(...)`)

Agree with Copilot's two comments here. Worth noting: with N concurrent misses on the same key, you get 1 real IdP call but N `cache.Put` calls and N "token exchange succeeded" log lines — and `TestExchange_ConcurrentMissesCollapse`'s own comment claims "ONE cache write," which isn't actually pinned by the test (it only asserts `callCount`, not write/log count). Under a real cold-start burst this multiplies log volume proportional to the stampede size, which undercuts part of the point of using singleflight here. One-line fix: capture `shared` and gate the Put+log on `!shared`.

**3. `GetResult`/`PutResult` don't carry `CachedCredential`'s log-safety with them** — `internal/oauth/cache/cache.go`

`CachedCredential` has `String()`/`LogValue()`/`GoString()`, but `GetResult`/`PutResult` embed it as a plain field without their own `LogValue()`. Confirmed `slog.Any("result", gr)` on a `GetResult` serializes via reflection and emits the raw token `Value` — bypassing both `CachedCredential`'s guards and `main.go`'s `ReplaceAttr` net (which only inspects top-level attr keys). Nothing logs the whole struct today, but it's exactly the landmine secure-logging-rules.md warns about — the next debug log line that does `slog.Any("cache_result", gr)` silently leaks a live broker-bound token. Worth adding `LogValue()` to both result types (exposing only `Status`) while this is fresh.

### Worth addressing, non-blocking

- **`factory.go` validates only `MaxSize`** (Copilot flagged this too): `MaxTTL <= 0` silently disables caching forever with no startup error. Low risk today given hardcoded defaults, but the PR description itself raises making these configurable via `broker-config.yaml` — validating now avoids a silent footgun later.
- **Cache hit/miss is invisible at the shipped default log level**: `GetHit`/`PutStored` map to Debug, and there's no hit-rate/size metric, so there's no operator-visible way to confirm the 4.5x speedup is actually happening in production, or to tell "cache working" from "cache misconfigured."
- **`exchanger_test.go`'s `mustTestCache()` is never closed** (Copilot flagged), used by ~40 call sites in that package — leaks Otter's background goroutines across the test suite.
- Minor, Copilot already has these right and I agree with the calls: `cache_test.go` bypassing `Close()` via type assertion instead of the interface method now that it exists, and the `main.go` cache-leak-on-`FromConfig`-error / dropped `Close()` error (both low severity — the leak path is immediately followed by `os.Exit(1)`, so it doesn't matter in practice, but worth a one-line comment so a future reviewer doesn't re-flag it).

Nice test coverage on the cache package and the exchanger↔cache contract otherwise — the hit/miss/expiry/singleflight-collapse tests are thorough and easy to follow.

**Comment by amitmorade** [comment]
Thanks Andrea — great pass. Rundown below, in your order.

## Worth fixing before merge

**1. `HandleAuthFailure` return-true regression** — you're right. Fixed in [`eff2080`](https://github.com/SolaceDev/solace-broker-mcp/pull/146/commits/eff2080): keeps the `Invalidate` (eviction helps the *next* request), returns `false` until SOL-151624 lands. Docstring rewritten, test updated. When SOL-151624 flips the resilience-layer wiring, this becomes a one-line change back to `true`.

**2. Singleflight redundant Put + logs** — already addressed, though not via the `!shared` flag. `Put` sits inside the singleflight func at `internal/tokenexchange/exchange.go:61-84`, so only the winner runs it — structurally impossible to Put more than once per burst regardless of caller behavior. Coverage: added `TestExchange_SingleflightWinnerOnlyWritesCache` (`exchange_test.go:1327`) which wraps the cache with a counter and asserts `puts == 1` under 100 concurrent goroutines. That's the test the older comment was implicitly claiming.

The one thing that stays N-times is the per-caller `"token exchange succeeded"` debug log outside the func — that's per-caller observability (correlation ID etc). Happy to reconsider if that's noise you'd rather not have.

**3. `GetResult`/`PutResult` slog reflection leak** — confirmed with a reproducer, `CachedCredential.LogValue` is bypassed because slog only consults the outermost type's guard. Fixed in [`f0d0ada`](https://github.com/SolaceDev/solace-broker-mcp/pull/146/commits/f0d0ada) with `LogValue` on both types + a regression test.

Sweeping for the same pattern found four *pre-existing* types with the identical landmine (`Token`, `ExchangeInput`, `Params`, `DeduplicationKeyInput`). Filed [SOL-151649](https://sol-jira.atlassian.net/browse/SOL-151649) rather than growing this PR — the two new types went with the PR that introduced them.

## Worth addressing, non-blocking

Three of four already in the branch:

- **`factory.go` validation** — `internal/oauth/cache/factory.go:14-25` validates `MaxTTL > 0` and `ClockSkew >= 0`.
- **`mustTestCache()` leak** — replaced with `validParams(t *testing.T)` at `exchanger_test.go:33`, wires `cachetest.Default(t)` + `t.Cleanup(Close)`.
- **`cache_test.go` type assertion + `main.go` leak-on-error** — `cache_test.go:22` uses the interface `Close()`; `main.go:472-482` closes on the `FromConfig` error path with the "we do this for the test that drives this path" comment.

**Hit/miss observability at default log level** — real gap, keeping out of this PR. Bumping `GetHit`/`PutStored` to INFO would produce one log line per tool call at scale. The right shape is a Prometheus hit-rate/size metric plus keeping the logs at Debug — its own design decision. Filing separately, will link when up.

## Other follow-ups filed from this review

Under parent epic [SOL-148417](https://sol-jira.atlassian.net/browse/SOL-148417):

- [SOL-151595](https://sol-jira.atlassian.net/browse/SOL-151595) — `AuthIdentity` value type for compile-time key safety (after scopes removal).
- [SOL-151600](https://sol-jira.atlassian.net/browse/SOL-151600) *(High)* — negative cache / backoff on IdP error.
- [SOL-151617](https://sol-jira.atlassian.net/browse/SOL-151617) — preserve tokens across caller-cancellation storms.
- [SOL-151649](https://sol-jira.atlassian.net/browse/SOL-151649) — extend `LogValue` to the four pre-existing types.
- SOL-151624 (pre-existing) — `PrepareRetry` wiring.

[SOL-151649]: https://sol-jira.atlassian.net/browse/SOL-151649?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Review by solace-aross — APPROVED** [review]
Verified the three blocking items against #150 HEAD (`f0d0ada`), not just the linked #146 commits. All landed.

1. **`HandleAuthFailure` return-true regression** — now returns `false` on both paths and keeps the `Invalidate`. No wasted round-trip and no misleading "recovered" WARN until SOL-151624 flips the resilience wiring. Correct call.

2. **Singleflight redundant Put + logs** — `Put` now sits inside `group.Do`, so the winner is the only writer. Structurally sound rather than flag-gated, and `TestExchange_SingleflightWinnerOnlyWritesCache` pins `puts == 1` under 100 goroutines. Better than the `!shared` approach I suggested. Agreed on leaving the per-caller "exchange succeeded" debug line outside the func; that's per-request observability, not noise.

3. **`GetResult`/`PutResult` reflection leak** — both now carry `LogValue()` exposing only `Status`, so `slog.Any` can't reflect `Entry.Value` into a record. Filing SOL-151649 for the four pre-existing types instead of growing this PR was the right scope call.

Non-blocking also confirmed in the branch: `factory.go` validates `MaxTTL > 0` and `ClockSkew >= 0`, `cachetest.Default` wires `t.Cleanup(Close)`, and `main.go` closes the cache on the `FromConfig` error path. Deferring hit-rate observability to a Prometheus metric rather than bumping the logs to INFO is the right shape.

Also checked myself: Close ordering in `main.go` is correct (the exchanger `Close` defer is registered before the server, so it runs after shutdown under LIFO), TTL math applies skew at write time, and the credential-safety guards hold. CI is green across build, lint, all three e2e suites, and FOSSA.

Approving. Good work on the honest known-limitation writeup and the follow-up tickets under the epic.
